src/fiber/graph.rs

```diff
@@ -1311,6 +1327,119 @@ where
         let penalty = default_attempt_cost * (1.0 / (0.5 - time_pref / 2.0) - 1.0);
         weight as u128 + (penalty / probability) as u128
     }
+
+    // This function is used to build the path from the specified path
+    // if there are multiple channels between the same two nodes and channel is not specified,
+    // we still try to select the best channel based on the channel's capacity and fee rate,
+    // there difference is that we don't need to use a aggregated distance which is used
+    // in the path finding algorithm, because we assume all the nodes in the path are
+    // already selected, and we only need to find the best channel for each hop.
+    pub(crate) fn build_path(
+        &self,
+        source: Pubkey,
+        command: BuildRouterCommand,
+    ) -> Result<Vec<RouterHop>, PathFindError> {
+        let mut router_hops = command.hops_info.clone();
+        router_hops.reverse();
+
+        let mut path = vec![];
+        // If not set, the minimum routable amount `1` is used
+        let mut agg_amount = command.amount.unwrap_or(1);
+        let mut agg_tlc_expiry = command
+            .final_tlc_expiry_delta
+            .unwrap_or(DEFAULT_TLC_EXPIRY_DELTA);
+        for current in 0..router_hops.len() {
+            let cur_hop = &router_hops[current];
+            let prev_hop = &router_hops.get(current + 1);
+            let prev_hop_pubkey = prev_hop.map(|x| x.pubkey).unwrap_or(source);
+            let mut found = None;
+            for (from, to, channel_info, channel_update) in self.get_node_inbounds(cur_hop.pubkey) {
+                if from != prev_hop_pubkey {
+                    continue;
+                }
+                if &command.udt_type_script != channel_info.udt_type_script() {
+                    continue;
+                }
+
+                // if specified channel outpoint is not empty, we will only use the specified channel
+                let channel_outpoint = channel_info.out_point().clone();
+                if cur_hop
+                    .channel_outpoint
+                    .clone()
+                    .unwrap_or(channel_outpoint.clone())
+                    != channel_outpoint
+                {
+                    continue;
+                }
+
+                let mut current_amount = agg_amount;
+                if current_amount > channel_info.capacity() {
```

Comment:
I think this check should be moved after forward fee calculation, could you add an unit test for this?

Reply:
Fixed with https://github.com/nervosnetwork/fiber/pull/577/commits/23f5925558d69b05a04492eb4e398be09e72d40c